Proposal for balance redesign

JGcarv opened the pull request:
Referencing #10 

Redesigned the balance tracking mechanism following the suggestion in the issue.

Ideally in the future would be good to support tracking token balances, as well sending tokens with the `send` module.

Let me know what you guys think!
Cheers,
JG

JGcarv commented:
I just made a new a new commit implemententing the comments:

* Brought `currentBalance` back to life
* Changed to `class` syntax
* Updated readme to use `const` and `should`

Let me know what you think about those!

frangio commented on src/balance.js:
```diff
@@ -1,17 +1,32 @@
 module.exports = {
-  current: balanceCurrent,
-  difference: balanceDifference,
+  balanceCurrent,
```
Following the convention that we use in other helpers (e.g. `time`), these two should be named `current` and `tracker`, so they can be used as `balance.current` and `balance.tracker`.

frangio commented on src/balance.js:
```diff
@@ -1,17 +1,32 @@
+  }
+  async get () {
+    const bal = new BN(await web3.eth.getBalance(this.account));
+    this.balances.push(bal);
```
Do we need to keep all past balances? It feels like keeping only the last one would be enough.

JGcarv commented:
Just addressed the other comments. 

I still have some mixed feeling about keeping only last two balances, since it adds some small complexity in the library, while the actual effect on the tests might be negligible, as I imagine it'll always be kept small enough to not affect performance.

But let me know what you guys think!

nventuro commented on src/balance.js:
```diff
@@ -1,17 +1,33 @@
+class Tracker {
+  constructor (acc) {
+    this.account = acc;
+    this.balances = [];
```
Using an array seems a bit overkill for this, since we only need to store the previous balance (to calculate `delta`). That way, we'd get rid of the `push`, `shift` and `length` calls, resulting in much simpler code. Something like:

```
get() {
  this.prev = current();
  return this.prev;
}

delta() {
  const current = current();
  const delta = current - this.prev;
  this.prev = current;
  return delta;
}
```

nventuro commented on test/src/balance.test.js:
```diff
@@ -11,17 +11,24 @@ contract('balance', function ([sender, receiver]) {
   });
 
-  describe('difference', function () {
+  describe('balance tracker', function () {
```
We should also add tests for `get`, and for how `get` and `delta` interact (i.e. the fact that `get` sets a new checkpoint).

JGcarv commented:
Thank you for the insightful review, @nventuro 

I removed the array completely, added a few more test cases and improves the documentation a bit. Let me know what you think!

nventuro commented on src/balance.js:
```diff
@@ -1,17 +1,33 @@
+
+async function balanceTracker (owner) {
+  const tracker = new Tracker(owner);
+  await tracker.get();
```
I think it might make more sense to put this inside of the `Tracker` constructor, making the class itself safer to use.

JGcarv replied:
You mean the first call to `get()`? Or export the whole classe and get rid of `balanceTracker` altogether?

nventuro replied:
Yes, the call to get. Exporting the class might make sense, it looks like the only difference would be having to use `new`, right? That may help people understand that the tracker is a proper object, but I'll leave it up to you.

JGcarv replied:
I think the main issue is that constructors cannot be `async`, so that would leave the call to `get()` hanging. Although it would be fine on almost all the cases, it's kind of fragile.The same issue happens when exporting the whole class, since it involves a async call to check the balance, the usage gets a bit weird, and inconsistent with `balance.current`.

One solution is to remove that first call completely and instruct users to always call `get()` before starting calling `delta()`, but I think it makes usage unnecessarily complicated.

Considering the trade offs, I would leave as is, but I'm not a heavy user of javascript so I might be missing some better solution.

nventuro replied:
Ah, I didn't remember about not being able to have `async` constructors. Let's leave it as is then, thanks!

nventuro commented on src/balance.js:
```diff
@@ -1,17 +1,33 @@
+class Tracker {
+  constructor (acc) {
+    this.account = acc;
+    this.current = async () => new BN(await web3.eth.getBalance(this.account));
```
Why not simply use `balanceCurrent`?

JGcarv replied:
I don't have a good reason, tbh. I think in my mental model `balance.current` and `balance.tracker` are two separate things and therefore it might make sense to make them independent. But I don't feel so strong about it though, so I'll just changing it!

JGcarv replied:
Done!

nventuro commented on test/src/balance.test.js:
```diff
@@ -11,17 +11,44 @@ contract('balance', function ([sender, receiver]) {
+  describe('balance tracker', function () {
+    it('returns current balance ', async function () {
+      const tracker = await balance.tracker(receiver);
+      (await tracker.get()).should.be.bignumber.equal(await web3.eth.getBalance(sender));
```
```suggestion
      (await tracker.get()).should.be.bignumber.equal(await web3.eth.getBalance(receiver));
```
